### ISSUE TITLE
Restarting Kernel in VS Code no longer removes hold on the DLL files

### ISSUE THREAD

**Issue by kmccalvin** [dotnet/interactive#2557]
### Describe the bug
I have a workflow where I use Visual Studio for writing code, and VS Code for writing C# Interactive code. 
In the past, I would run some code snippet in Polyglot Notebook, click "Restart" (kernel), and then make changes in Visual Studio and Rebuild the projects. I needed to click "Restart" because otherwise Visual Studio is unable to move/modify the dll files that Polyglot Notebooks is holding onto. 

However, recently the new version of the **Polyglot Notebooks** extension forces me to install .NET 7 SDK. After doing so, my workflow no longer works. More specifically, clicking "Restart" now no longer enables Visual Studio to rebuild my projects. Restarting VS code works but isn't an ideal workflow for me. 

I have now downgraded to version v1.0.3557011 (randomly chosen) and uninstalled .NET 7 SDK to re-enable my workflow. 

### Please complete the following:

**Which version of .NET Interactive are you using?** (In a notebook, run the `#!about` magic command. ):

- OS: Windows 11
- Frontend: Visual Studio Code with Polyglot Notebooks

Working version:
```
Version: 1.0.355701+da7df87cf78413155f55bbfe16c0c2a82c79162e
Library version: 1.0.0-beta.22557.1+da7df87cf78413155f55bbfe16c0c2a82c79162e
Build date: 2022-12-08T19:11:38.1859027Z
```
Breaking version:
```
Version: 1.0.360602+9bf026dabac44a6256a65168fa93dcb7e2edcca4
Library version: 1.0.0-beta.22606.2+9bf026dabac44a6256a65168fa93dcb7e2edcca4
Build date: 2022-12-08T18:49:35.7173095Z
```

### Screenshots
Here's the log file for failed attempt to rebuild solution
```
4>------ Rebuild All started: Project: GatewayUnitTests, Configuration: Debug Any CPU ------
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5599,5): warning MSB3061: Unable to delete file "C:\[Redacted]\Notebook\bin\Debug\net6.0\Notebook.dll". Access to the path 'C:\[Redacted]\Notebook\bin\Debug\net6.0\Notebook.dll' is denied.
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5599,5): warning MSB3061: Unable to delete file "C:\[Redacted]\Notebook\bin\Debug\net6.0\AnalyticalBalance.dll". Access to the path 'C:\[Redacted]\Notebook\bin\Debug\net6.0\AnalyticalBalance.dll' is denied.
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5599,5): warning MSB3061: Unable to delete file "C:\[Redacted]\Notebook\bin\Debug\net6.0\Gateway.dll". Access to the path 'C:\[Redacted]\Notebook\bin\Debug\net6.0\Gateway.dll' is denied.
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 2 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 2 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 3 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 3 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 4 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 4 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 5 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 5 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 6 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 6 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 7 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 7 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 8 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 8 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 9 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 9 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Beginning retry 10 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): warning MSB3026: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Beginning retry 10 in 1000ms. The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): error MSB3027: Could not copy "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". Exceeded retry count of 10. Failed. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): error MSB3027: Could not copy "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". Exceeded retry count of 10. Failed. The file is locked by: ".NET Host (10832)"
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): error MSB3021: Unable to copy file "C:\[Redacted]\Gateway\bin\Debug\net6.0\Gateway.dll" to "bin\Debug\net6.0\Gateway.dll". The process cannot access the file 'bin\Debug\net6.0\Gateway.dll' because it is being used by another process.
3>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4809,5): error MSB3021: Unable to copy file "C:\[Redacted]\AnalyticalBalance\bin\Debug\net6.0\AnalyticalBalance.dll" to "bin\Debug\net6.0\AnalyticalBalance.dll". The process cannot access the file 'bin\Debug\net6.0\AnalyticalBalance.dll' because it is being used by another process.
3>Done building project "Notebook.csproj" -- FAILED.
4>GatewayUnitTests -> C:\[Redacted]\GatewayUnitTests\bin\Debug\net6.0\GatewayUnitTests.dll
========== Rebuild All: 3 succeeded, 1 failed, 0 skipped ==========
```

**Comment by brettfo** [comment]
Do you have an example project you can share that shows this behavior?

**Comment by pgrit** [comment]
I have the same issue. Steps to reproduce:
1. Create a new project with any template, e.g.,
```
mkdir Test && cd Test
dotnet new classlib && dotnet build
```
2. Create a notebook, open it in VS Code, and add the following (seems to be important that you actually use a type from the assembly):
```csharp
#r "bin/Debug/net7.0/Test.dll"
Test.Class1 t = new();
```
3. Restart the notbook's kernel
4. Do a change in the project that requires a rebuild. E.g., add a property to the auto generated class in the classlib template:
```csharp
namespace Test;
public class Class1
{
    public int Foo { get; set; }
}
```
5. Rebuild via `dotnet build` to trigger the above issue

**Comment by kmccalvin** [comment]
Apologies for the late response. 
I have created a simple project in a new repository: https://github.com/kmccalvin/barecsharp
Steps to recreate:
 - Checkout repository
 - Build project with `dotnet build`
 - Run notebook code
 - Attempt to rebuild project with `dotnet build --no-incremental`

Here's a screenshot. 
![image](https://user-images.githubusercontent.com/120133088/211320455-0b27c2dd-76fd-40bf-a424-ea2c54a2d085.png)

**Comment by marckruzik** [comment]
I have the same problem. My current workaround is to kill all ".NET Host" processes.
There usually 4 to 8 of them, and killing one usually kills another one.
I am giving two ways to kill those processes, which allows to build a new dll on Visual Studio, and not relaunch VS Code.

__Kill ".NET Host" processes with the Task Manager__
1. open the Task Manager
2. order the processes by name
3. select a ".NET Host" process
4. kill the process by clicking on "end task" or the delete key
5. go back to step 3 until there are no more ".Net Host" processes (sometimes processes can come back, so focus on killing the bigger ones)
6. you can build the dll
7. re-execute the notebook from the beginning (ny killing the .NET Host processes you killed the kernel)

__Kill ".NET Host" processes with the command line__
1. open the command line
2. type `taskkill /IM "dotnet.exe" /F`
3. all ".NET Hosts" processes are killed at the same time
4. you can build the dll
5. re-execute the notebook from the beginning (ny killing the .NET Host processes you killed the kernel)

Here is an example:

![2023_01_22-18h07m58s-End_Folder_Purity - Microsoft Visual Studio](https://user-images.githubusercontent.com/51711029/213929946-c2a66615-3702-4712-aeac-24108405d4f0.png)

* Visual Studio could not copy the dll
* I used the command line and it killed all the .NET Host processes
* Visual Studio can copy the dll

**Comment by brettfo** [comment]
We just pushed (less than an hour ago) an update to VS Code - Insiders that may have a fix for this, can you try it again?

**Comment by kmccalvin** [comment]
> We just pushed (less than an hour ago) an update to VS Code - Insiders that may have a fix for this, can you try it again?

Sorry, I'm not up familiar with the verbiage. Is this something I can test? 
In VSCode, the latest Polyglot Notebook version I see is still from a month ago. 

![image](https://user-images.githubusercontent.com/120133088/214860208-58740c4e-eb78-46a0-8def-01efdf0fc7df.png)

**Comment by pgrit** [comment]
@brettfo 
> We just pushed (less than an hour ago) an update to VS Code - Insiders that may have a fix for this, can you try it again?

Works like a charm for me, thanks :)

@kmccalvin

> Sorry, I'm not up familiar with the verbiage. Is this something I can test?
In VSCode, the latest Polyglot Notebook version I see is still from a month ago.

https://code.visualstudio.com/insiders/

**Comment by kmccalvin** [comment]
@pgrit  
Thank you, I have downloaded insiders and tested it out.  
@brettfo  
Thanks! It now works for me. Restarting Kernel now allows me to re-build.  

Also unrelated: I noticed the hotkeys are working again. (`Ctrl + /` comments a line)  
I presume that's another issue being addressed live in the insiders version.

**Comment by brettfo** [comment]
@pgrit @kmccalvin Glad to hear it's working.  As for the comment/un-comment feature, yes, that's also something that we recently fixed in the insiders build in #2592.

In the first week (-ish) of each month we ship a new version of the extension to stable VS Code (i.e., the normal, not insiders build), so you'll soon see these fixes there, too.

All of these fixes will make it to the stable build of VS Code when the next version ships out, usually in the first week or so of the next month.
